contrib/dracut/90zfs/module-setup.sh.in

```diff
@@ -20,59 +20,30 @@ depends() {
 
 installkernel() {
 	instmods zfs
-	instmods zcommon
-	instmods znvpair
-	instmods zavl
-	instmods zunicode
-	instmods zlua
-	instmods icp
-	instmods spl
-	instmods zlib_deflate
-	instmods zlib_inflate
 }
 
 install() {
-	inst_rules @udevruledir@/90-zfs.rules
-	inst_rules @udevruledir@/69-vdev.rules
-	inst_rules @udevruledir@/60-zvol.rules
-	dracut_install hostid
-	dracut_install grep
-	dracut_install @sbindir@/zgenhostid
-	dracut_install @sbindir@/zfs
-	dracut_install @sbindir@/zpool
-	# Workaround for https://github.com/openzfs/zfs/issues/4749 by
-	# ensuring libgcc_s.so(.1) is included
-	if ldd @sbindir@/zpool | grep -qF 'libgcc_s.so'; then
-		# Dracut will have already tracked and included it
-		:;
-	elif command -v gcc-config >/dev/null 2>&1; then
-		# On systems with gcc-config (Gentoo, Funtoo, etc.):
-		# Use the current profile to resolve the appropriate path
-		s="$(gcc-config -c)"
-		dracut_install "/usr/lib/gcc/${s%-*}/${s##*-}/libgcc_s.so"*
-	elif [ "$(echo /usr/lib/libgcc_s.so*)" != "/usr/lib/libgcc_s.so*" ]; then
-		# Try a simple path first
-		dracut_install /usr/lib/libgcc_s.so*
-	elif [ "$(echo /lib*/libgcc_s.so*)" != "/lib*/libgcc_s.so*" ]; then
-		# SUSE
-		dracut_install /lib*/libgcc_s.so*
-	else
-		# Fallback: Guess the path and include all matches
-		dracut_install /usr/lib*/gcc/**/libgcc_s.so*
-	fi
-	# shellcheck disable=SC2050
-	if [ @LIBFETCH_DYNAMIC@ -gt 0 ]; then
-		for d in $libdirs; do
-			[ -e "$d/@LIBFETCH_SONAME@" ] && dracut_install "$d/@LIBFETCH_SONAME@"
-		done
-	fi
-	dracut_install @mounthelperdir@/mount.zfs
-	dracut_install @udevdir@/vdev_id
-	dracut_install awk
-	dracut_install cut
-	dracut_install tr
-	dracut_install head
-	dracut_install @udevdir@/zvol_id
+	inst_rules \
+		@udevruledir@/90-zfs.rules \
+		@udevruledir@/69-vdev.rules \
+		@udevruledir@/60-zvol.rules
+	
+	dracut_install \
+		@sbindir@/zgenhostid \
+		@sbindir@/zfs \
+		@sbindir@/zpool \
+		@udevdir@/vdev_id \
+		@udevdir@/zvol_id \
+		@mounthelperdir@/mount.zfs \
+		hostid \
+		grep \
+		awk \
+		tr \
+		cut \
+		head
+		
+	inst_libdir_file "libgcc_s.so*"
```

Comment:
This discards a bunch of custom logic, especially wrt Gentoo, that I believe is still relevant. I don't think `inst_libdir_file` properly searches the `/usr/lib*/gcc` hierarchy, for example. You probably don't want to abandon the sophisticated search that is currently done for this library.

Reply:
The logic there was failing on Debian Bullseye, which is why I migrated it to this (which absolutely works). I do not have a gentoo machine to test this on. If it can be tested by someone on gentoo that would be very nice.

As an aside, the initramfs was perfectly functional without this library as well.

Reply:
Of course it works on Debian, because Debian puts the library where `inst_libdir_file` can find it. That does not appear to be true on Gentoo (or at least *was* not true, hence the Gentoo-specific clause). I don't use Gentoo and can't confirm that the situation hasn't changed, but we at least know that removing the specific checks here will break whatever systems the checks were added to fix.

ZFSBootMenu had an issue with [`libgcc_s` in OpenSUSE](https://github.com/zbm-dev/zfsbootmenu/issues/211) and solved the problem by adopting `inst_libdir_file` to broaden its search. However, [we solved this the right way](https://github.com/openzfs/zfs/pull/13010) by keeping all of the special-purpose logic but trying to use `inst_libdir_file` in the generic case. The last-ditch 'just try anything' is also preserved. Also note that the if-else ladder in the `zfs` module has been replaced by a series of nested ifs, because it's better to fall through to more generic attempts to find the library even on those systems where the specific options would be tried first.

On something as critical as the filesystem module that is required to boot a system, dropping specific logic just because it isn't needed on your system is a very risk proposition.

Reply:
The point I was trying to make was that the special purpose logic here _did not work_ for Debian Bullseye. There is a case it misses. Fortunately, it seems that `zpool` and `zfs` on Debian do not link to `libgcc_s.so.1`.

I think the best way of replacing that entire logic would be with:

```sh
local libgcc="$(ldd /sbin/zpool | awk '/libgcc_s.so/ {print $3}')"

if [[ -f $libgcc ]]; then
    dracut_install $libgcc
fi
```

What do you think?

Reply:
I understand the old method broke under Debian, which is why the `inst_libdir_file` is better than a blind copy. We still need the custom logic to work around unexpected breakage elsewhere. Grepping through ldd output is not sufficient of appropriate because dracut already does that in the installation function. The special cases are to handle missed dependencies, probably because some distributions wind up `dlopen`-ing the library.

Reply:
Makes sense. I'll make the changes accordingly and push them when I get back to my workstation.